### PR TITLE
Add consensus check on transferdomain accounting

### PR DESCRIPTION
**Pull request by Mixa84**
## Summary

- Tie in accounting of transferdomain into consensus.
- Check old values of DvmIn, DvmOut, DvmCurrent, EvmIn, EvmOut and EvmCurrent with all block transactions processed and if sums match.

## Implications

- Storage
  - [ ] Database reindex required
  - [ ] Database reindex optional
  - [ ] Database reindex not required
  - [ ] None

- Consensus
  - [X] Network upgrade required
  - [ ] Includes backward compatible changes
  - [ ] Includes consensus workarounds
  - [X] Includes consensus refactors
  - [ ] None

### CODE REVIEW

**Comment by prasannavl** [comment]
TODO: 

- Restore previous checks
- Add EVM tally

File: lib/ain-rs-exports/src/evm.rs [hunk 1]
```diff
@@ -667,6 +709,35 @@ pub fn evm_try_is_dst20_deployed_or_queued(
...
+    token_id: u64,
+    state_root: &str,
+) -> u64 {
+    let state_root = match state_root {
+        "" => None,
+        _ => match state_root.parse() {
+            Ok(state_root) => Some(state_root),
+            Err(_) => return cross_boundary_error_return(result, "Invalid state root"),
+        },
+    };
```
Comment (shohamc1) [inline lib/ain-rs-exports/src/evm.rs:723]:
```suggestion
    let Ok(state_root) = state_root.parse() else {
        return cross_boundary_error_return(result, "Invalid state root");
    };
```

File: src/masternodes/validation.cpp [hunk 2]
```diff
@@ -2400,7 +2402,182 @@ static Res ValidateCoinbaseXVMOutput(const XVM &xvm, const FinalizeBlockCompleti
...
+    CustomTxType txType = GuessCustomTxType(*tx, metadata);
+    if (txType == CustomTxType::TransferDomain) {
+        auto txMessage = customTypeToMessage(txType);
+        if (CustomMetadataParse(pindex->nHeight, chainparams.GetConsensus(), metadata, txMessage))
+            return *(std::get_if<CTransferDomainMessage>(&txMessage));
+    }
+    return CTransferDomainMessage{};
+}
+
+CAmount GetEvmDST20TotalSupply(const DCT_ID& id, const std::string &stateRoot = std::string()) {
```
Comment (shohamc1) [inline src/masternodes/validation.cpp:2416]: Would it be safer to have this as `ResVal<CAmount>`?

File: src/masternodes/validation.cpp [hunk 3]
```diff
@@ -2400,7 +2402,182 @@ static Res ValidateCoinbaseXVMOutput(const XVM &xvm, const FinalizeBlockCompleti
...
+    std::map<CScript, CStatsTokenBalances> &evmBalances = evmInitialState.evmBalances;
+    CStatsTokenBalances& dst20EvmTotalSupply = evmInitialState.dst20EvmTotalSupply;
+
+    // Storing inital EVM address balances
+    for (const auto &tx : block.vtx) {
+        auto txMessage = DecodeTransferDomainMessage(tx, pindex, chainparams);
+        for (auto const &[src, dst]: txMessage.transfers){
+            if (src.amount.nTokenId == DCT_ID{0}) {
+                std::vector<std::tuple<const uint8_t, VMDomain, const CScript&, const CTokenAmount&, bool>> balanceList{
+                    { src.domain, VMDomain::EVM, src.address, src.amount, evmBalances.find(src.address) == evmBalances.end() },
```
Comment (shohamc1) [inline src/masternodes/validation.cpp:2448]: Doesn't seem necessary to track `src.amount` here.

Comment (Mixa84) [reply src/masternodes/validation.cpp:2448]: Simplified a bit

File: src/masternodes/validation.cpp (hunk 2 above)
Comment (Mixa84) [reply src/masternodes/validation.cpp:2416]: Correct, switched to ResVal. Thnx

File: lib/ain-rs-exports/src/evm.rs (hunk 1 above)
Comment (Mixa84) [reply lib/ain-rs-exports/src/evm.rs:723]: Couldn't make it compile this way...

**Review by Jouzo — APPROVED** [review]

**Review by Bushstar — APPROVED** [review]

**Pull request merged**